nexus-and-alo/nexus-strawman.md

```diff
@@ -0,0 +1,361 @@
+
+**Nexus Worker:** A special purpose Worker that pulls Nexus Tasks for its registered Nexus Instances from the Server and executes them using associated Nexus Task Handlers. 
+
+## User flows
+
+### Hosting a Nexus Service
+
+1. User instantiates Nexus Service Instance using provided `nexus` CLI commands (also possible via SDK or API).
+1. User defines a Nexus Task Handler that will handle the processing of Nexus Tasks for the instantiated Nexus Service Instance.
+1. User hosts their Nexus Task Handler using a Nexus Worker. They must specify the Nexus Service Instance ID when starting the Nexus Worker.
+
+	// TODO (Chad): show how registration would work for generic handler
+    ```go
+    func main() {
+        ctx := context.TODO()
+
+        // Create Nexus client
+        client, err := client.Dial(ctx, client.Options{Target: "nexus-backend.example.com"})
+        if err != nil {
+            log.Fatal(err)
+        }
+        defer client.Close()
+
+        // Build worker with Temporal pieces
+        temporalClient, err := temporalclient.Dial(temporalclient.Options{HostPort: "nexus-backend.example.com:7233"})
+        if err != nil {
+            log.Fatal(err)
+        }
+        defer temporalClient.Close()
+        builder, err := temporalnexus.NewWorkerBuilder(temporalnexus.WorkerBuilderOptions{Client: temporalClient})
+        if err != nil {
+            log.Fatal(err)
+        }
+        // Add workflow as ALO to Nexus
+        err = builder.AddWorkflowALOHandler("foo", temporalnexus.WorkflowALOHandlerOptions{
+            Workflow: FooWorkflow,
+            Options:  temporalclient.StartWorkflowOptions{TaskQueue: "my-task-queue"},
+        })
+        if err != nil {
+            log.Fatal(err)
+        }
+
+        // Worker definition...
+    }
+    ```
+1. Nexus is now able to receive and process traffic on its frontend endpoint.
+
+### Consuming a Nexus Service from Temporal
+
+1. User asks the Nexus Service author for the endpoint of the Nexus Service Instance they are trying to consume. For Temporal based Nexus Service Instances, the endpoint will not be HTTP or gRPC but rather a custom uri formulation.
+2. User registers provided endpoint to Nexus Worker, similarly to how you would register Workflows and Activities with a Worker. 
+3. User calls the external Nexus Service using a Workflow-specific Nexus API which is provided by Temporal.
+    ```go
+    // calling a Nexus from a Temporal Workflow
+    func MyWorkflow(ctx workflow.Context, name string) (string, error) {
+        client := temporalnexus.NewWorkflowClient(temporalnexus.WorkflowClientOptions{Endpoint: "temporal:72aA64A0jhOL99B2"})
+        // Start some alo
+        alo, err := client.StartALO(ctx, temporalnexus.WorkflowStartALOOptions{
+            Name: "foo",
+            ID:   "my-id",
```

Comment:
Feel free to resolve if too detailed: It may make a lot of sense to only allow users to specify a "human name" (which need not be unique) and have the client always use a UUID for the real ID, and return that.

Reply:
Yes, completely. It just avoids a footgun where the user doesn't actually use unique IDs. The call returns the UUID, which they then persist. The passed-in ID may be non unique (though typically, will also be unique) and serves as a human readable alternative.

Reply:
The ID should be unique globally (only in the target Nexus?). It's used as an idempotency key, for workflow based ALOs we'll use this ID as the workflow ID.
@mfateev mentioned at the conference that in a workflow we can default to a deterministic UUID, I agree, we do the same for child workflows now.

Reply:
I tend to side with Roey and Chad on this one. What I proposed to Max is that we always require client side ID, but in the case of calling an ALO from a Workflow, we could automatically handle the ID generation state-fully and therefore not explicitly require it from the caller.